draft-ietf-quic-transport.md

```diff
@@ -2613,6 +2627,14 @@ The NEW_CONNECTION_ID is as follows:
 
 The fields are:
 
+Sequence:
+
+: A variable-length integer.  This value starts at 0 and increases by 1 for each
+  connection ID that is provided by the server.  The connection ID that is
+  assigned during the handshake is assumed to have a sequence of -1.  That is,
```

Comment:
Is there a real reason we can consider the handshake CID `0` and start from there? I'd rather not have to deal with negative numbers in my code.

Reply:
Probably not.  This is just the existing text.  It's actually a little worse than that, because the Server's Preferred Address gives the client a CID to use when probing, so there can actually be two CIDs prior to the first one issued by frame.

Reply:
Then maybe we define handshake as `0`, the server's preferred address as `1` (even if not present) and then new CIDs start from there (i.e. `2`). I don't see any reason to prevent gaps in the CID sequence number space either.

Reply:
The alternative being -2, -1, and starting from 0.  I think the logic there is the principle that we should try to avoid there being invalid values, and 0/1 would be invalid values still present in the frame.  If the frame can carry a 0, then 0 should be the first value value.

Reply:
This also creates an odd situation in that SPA prompts the client to probe, but the server won't have a corresponding CID at that sequence number, so it will have to use the first CID the client issued to it -- *if* the client issues one with/before the probe! -- which will force the client to advance again.

Reply:
There can be very many connection IDs during the handshake with this new Retry packet, so it's all special casing here.  Start at 2^16-N if you feel compelled to track things this way.

Reply:
What if we just eliminate this statement and let it be implementation-local?  The new text I added to the CID section says that "The series of
connection IDs issued by an endpoint is ordered, with the final connection ID
selected during the handshake coming first."

We could also fix this by removing the CID from SPA and simply saying that the server needs to send the client a CID it can use for the probe pretty promptly if it wants the client to do so.

Reply:
I don't think that would help with the suggestion in #1468.  Insisting on a fixed starting point ensures that you can do things like matching.

Removing the connection ID from the preferred address would make it harder to get that right because you then need to wait for a NEW_CONNECTION_ID frame before migrating.

Reply:
I've merged this, so let's open an issue to close this out.